flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/declarative/ExecutingTest.java

```diff
@@ -0,0 +1,451 @@
+            // transition EG into terminal state, which will notify the Executing state about the
+            // failure (async via the supplied executor)
+            exec.getExecutionGraph().failJob(new RuntimeException("test failure"));
+        }
+    }
+
+    @Test
+    public void testTransitionToFinishedOnSuspend() throws Exception {
+        try (MockExecutingContext ctx = new MockExecutingContext()) {
+            Executing exec = getExecutingState(ctx);
+            ctx.setExpectedStateChecker((state) -> state == exec);
+
+            ctx.setExpectFinished(
+                    archivedExecutionGraph -> {
+                        assertThat(archivedExecutionGraph.getState(), is(JobStatus.SUSPENDED));
+                    });
+            exec.onEnter();
+            exec.suspend(new RuntimeException("suspend"));
+        }
+    }
+
+    @Test
+    public void testScaleUp() throws Exception {
+        try (MockExecutingContext ctx = new MockExecutingContext()) {
+            Executing exec = getExecutingState(ctx);
+            exec.onEnter();
+
+            ctx.setExpectRestarting(
+                    restartingArguments -> {
+                        // expect immediate restart on scale up
+                        assertThat(restartingArguments.getBackoffTime(), is(Duration.ZERO));
+                    });
+            ctx.setCanScaleUp(() -> true);
+            exec.notifyNewResourcesAvailable();
+        }
+    }
+
+    @Test
+    public void testNoScaleUp() throws Exception {
+        try (MockExecutingContext ctx = new MockExecutingContext()) {
+            Executing exec = getExecutingState(ctx);
+            ctx.setCanScaleUp(() -> false);
+            exec.onEnter();
+            exec.notifyNewResourcesAvailable();
+        }
+    }
+
+    @Test
+    public void testFailingOnDeploymentFailure() throws Exception {
+        try (MockExecutingContext ctx = new MockExecutingContext()) {
+            ctx.setCanScaleUp(() -> false);
+            ctx.setHowToHandleFailure(Executing.FailureResult::canNotRestart);
+            ctx.setExpectFailing(assertNonNull());
+
+            // create ExecutionGraph with one ExecutionVertex, which fails during deployment.
+            JobGraph jobGraph = new JobGraph(new JobVertex("test"));
+            Executing exec = getExecutingState(ctx, null, jobGraph);
+            TestingLogicalSlotBuilder slotBuilder = new TestingLogicalSlotBuilder();
+            TaskManagerGateway taskManagerGateway =
+                    new ExecutionVertexDeploymentTest.SubmitFailingSimpleAckingTaskManagerGateway();
+            slotBuilder.setTaskManagerGateway(taskManagerGateway);
+            LogicalSlot slot = slotBuilder.createTestingLogicalSlot();
+            exec.getExecutionGraph()
+                    .getAllExecutionVertices()
+                    .forEach(executionVertex -> executionVertex.tryAssignResource(slot));
+
+            // trigger deployment
+            exec.onEnter();
+        }
+    }
```

Comment:
We are missing the test that a failing deployment goes to restarting if we can restart.

Reply:
I believe the test above (and the additional test you are proposing) are not needed, as it is not the responsibility of the Executing state to handle failures during deployment. Of course it is, but it is indirect. Failures during deployment are reported as `updateTaskExecutionState()` or `handleGlobalFailure` via the `UpdateSchedulerNgOnInternalFailuresListener`, which should be tested separately (as part of tests for the declarative scheduler itself).

Reply:
If the `Executing` state no longer implements `InternalFailuresListener`, then this should be the best solution.